### PR TITLE
Rewrite son to work based on generators

### PR DESCRIPTION
**Pull request by sirmarcel**
This is neccessary to be able to parse files that are large. Using generators means that we don't need to first load the full file into memory, we do everything on the fly.

The outward-facing interface of `son.load` and `son.dump` remains unchanged, but we add `son.read`, which returns an iterator over entries in the file.

In order to do this cleanly I rewrote almost everything, except the tests. ;) 

I think the reorganised code is mostly self-documenting, but happy to discuss any particular changes. Do you agree with the naming of `son.read`? I also considered `son.open` in line with `xarray`, but `open` is a reserved name in `python`, so it's awkward to work with.

### CODE REVIEW

**Comment by sirmarcel** [comment]
I'm also considering changing `read` to return `(metadata, iterator_over_non_metadata)` instead of just the `iterator`. This makes its logic slightly more complex, but it might be worth the trade-off for ease of use. Otherwise, the logic of separating the (potential) first metadata entry from the rest falls to client code, which is probably a bit annoying.

**Comment by flokno** [comment]
Regarding the name: I think `open` instead of `read` would be more in line not only with `xarray`, but also with `Path`. `read` has the connotation of "read everything" to me, like `Path.read_text`. So yes I would prefer `open`.

File: son/interface.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def dump(obj, file, is_metadata=False, **kwargs):
+    """dump an object to son file, metadata can only be written once
+
+    Args:
+        obj (Object): object to be dumped to string
+        file (str or Path): file to dump to
+        is_metadata (bool): is this a metadata entry?
+        **kwargs: args to be passed to son.dumps
```
Comment (flokno) [inline son/interface.py:21]: `son.dumps` doesn't exist anymore, right?

File: son/interface.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        msg = "{} exists, possibility of data loss!".format(file)
+        raise FileExistsError(msg)
+
+    string = serialize.dump(obj, **kwargs)
+
+    with open(file, "a") as f:
+        stream.write(f, string, is_metadata=is_metadata)
+
+
+def read(file, verbose=False, encoding="utf-8", **kwargs):
```
Comment (flokno) [inline son/interface.py:34]:
```suggestion
def open(file, verbose=False, encoding="utf-8", **kwargs):
```

as discussed

File: son/serialize.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+import warnings
+
+try:
+    import ujson as json
+except ModuleNotFoundError:
+    import json
+
+
+def dump(obj, dumper=json.dumps, **kwargs):
```
Comment (flokno) [inline son/serialize.py:18]: I wonder if we want to keep the name `dumper`, it sound .. dumb.

**Review by flokno — APPROVED** [review]
Looks good :)

File: son/serialize.py (hunk 3 above)
Comment (sirmarcel) [reply son/serialize.py:18]: I think it's fine, it mirrors the language of `json`. Don't see a reason to break the API over this.

File: son/interface.py (hunk 2 above)
Comment (sirmarcel) [reply son/interface.py:34]: Ok

File: son/interface.py (hunk 1 above)
Comment (sirmarcel) [reply son/interface.py:21]: it should be `son.serialize.dump`, yes. will fix

**Comment by flokno** [comment]
Thank you very much!

**Pull request merged**